File: leetcode/cpp/zero-array-transformation-i.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        vector<int> freq(n+1,0);
        for(auto q : queries){
            auto l = q[0],r=q[1];
            freq[l] = 1;
            freq[r+1] = -1;
        }
        for(int i = 1; i <n; i++){
            freq[i] += freq[i-1];
        }
        for(int i = 0; i < n; i++){
            if(nums[i]>freq[i]){
                return false;
            }
        }
        return true;
    }
};

File: leetcode/java/ZeroArrayTransformationI.java
public class ZeroArrayTransformationI {
    public boolean isZeroArray(int[] nums, int[][] queries) {
        int n = nums.length;
        int[] prefix = new int[n];
        for(int[] query : queries){
            int l = query[0], r = query[1];
            prefix[l] -= 1;
            if(r+1 < n) prefix[r+1] += 1;
        }
        for(int i = 1; i < n; i++){
            prefix[i]+=prefix[i-1];
        }
        for(int i = 0; i < n; i++){
            if(nums[i] + prefix[i] > 0) return false;
        }
        return true;
    }
}